File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/find_all_numbers_disappeared_in_an_array/FindAllNumbersDisappearedinanArray.java
package main.java.com.jiang.leetcode100.find_all_numbers_disappeared_in_an_array;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] != i + 1) {
                int t = nums[i];
                if(nums[t - 1] != t) {
                    nums[i] = nums[t - 1];
                    nums[t - 1] = t;
                    i--;
                }
            }
        }
        List<Integer> result = new ArrayList<>();
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] != i + 1) result.add(i + 1);
        }
        return result;
    }
}

public class FindAllNumbersDisappearedinanArray {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums.length; i++) nums[i] = Integer.parseInt(nums_str[i]);

        List<Integer> result = new Solution().findDisappearedNumbers(nums);
        for(Integer num : result) System.out.print(num + " ");
    }
}

File: Leetcode100/leetcode100_python/FindAllNumbersDisappearedinanArray.py
from typing import List

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        for i in range(len(nums)):
            if nums[i] != i + 1:
                while nums[i] != nums[nums[i] - 1]:
                    t = nums[i]
                    nums[i], nums[t - 1] = nums[t - 1], nums[i]
        result = []
        for i in range(len(nums)):
            if nums[i] != i + 1:
                result.append(i + 1)
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findDisappearedNumbers(nums))